### ISSUE TITLE
enhancement: axis argument to transforms

### ISSUE THREAD

**Issue by aaren** [PyWavelets/pywt#115]
At present, transforms can only be performed on arrays of the corresponding dimensionality. It would be useful to introduce an `axis` argument to allow the transform to be computed over arbitrary array dimensions, e.g.

```python
a = np.ones((3, 4))
pywt.dwt(a, wavelet='db3', axis=-1)
```

This can be trivially implemented for `dwt` by using `pywt._pywt.dwt_axis` (introduced in #104), but will require some thought for higher dimensions.

The API would be the same as that for `np.fft.fft2` and `np.fft.fftn`:

```markdown
Signature: np.fft.fft2(a, s=None, axes=(-2, -1))
Docstring:
Compute the 2-dimensional discrete Fourier Transform

This function computes the *n*-dimensional discrete Fourier Transform
over any axes in an *M*-dimensional array by means of the
Fast Fourier Transform (FFT).  By default, the transform is computed over
the last two axes of the input array, i.e., a 2-dimensional FFT.

Parameters
----------
a : array_like
    Input array, can be complex
s : sequence of ints, optional
    Shape (length of each transformed axis) of the output
    (`s[0]` refers to axis 0, `s[1]` to axis 1, etc.).
    This corresponds to `n` for `fft(x, n)`.
    Along each axis, if the given shape is smaller than that of the input,
    the input is cropped.  If it is larger, the input is padded with zeros.
    if `s` is not given, the shape of the input along the axes specified
    by `axes` is used.
axes : sequence of ints, optional
    Axes over which to compute the FFT.  If not given, the last two
    axes are used.  A repeated index in `axes` means the transform over
    that axis is performed multiple times.  A one-element sequence means
    that a one-dimensional FFT is performed.
```

```markdown
Signature: np.fft.fftn(a, s=None, axes=None)
Docstring:
Compute the N-dimensional discrete Fourier Transform.

This function computes the *N*-dimensional discrete Fourier Transform over
any number of axes in an *M*-dimensional array by means of the Fast Fourier
Transform (FFT).

Parameters
----------
a : array_like
    Input array, can be complex.
s : sequence of ints, optional
    Shape (length of each transformed axis) of the output
    (`s[0]` refers to axis 0, `s[1]` to axis 1, etc.).
    This corresponds to `n` for `fft(x, n)`.
    Along any axis, if the given shape is smaller than that of the input,
    the input is cropped.  If it is larger, the input is padded with zeros.
    if `s` is not given, the shape of the input along the axes specified
    by `axes` is used.
axes : sequence of ints, optional
    Axes over which to compute the FFT.  If not given, the last ``len(s)``
    axes are used, or all axes if `s` is also not specified.
    Repeated indices in `axes` means that the transform over that axis is
    performed multiple times.
```

I don't know how useful this would be but it seems a natural step in extending the API. I'm not sure what exactly you'd do with the higher dimensional stuff but I have used a one-dimensional (continuous) wavelet transform over 3D data before.

**Comment by rgommers** [comment]
Makes sense to me.

**Comment by aaren** [comment]
Closed by #116 and #130.
